### PR TITLE
TRUNK-5352:Create HibernateDiagnosisDAO implementation of the DiagnosisDAO (Work In Progress)

### PR DESCRIPTION
**Pull request by CeciliaCaroline**
## Description of what I changed
Create HibernateDiagnosisDAO implementation of the DiagnosisDAO

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5352

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
+ * @since 2.2
+ */
+
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Can you remove the above empty line?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+public class HibernateDiagnosisDAO implements DiagnosisDAO {
+	
+	protected static final Log log = LogFactory.getLog(DiagnosisDAO.class);
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Make the above private

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	 * @param diagnosisId the id of the diagnosis to look for
+	 * @return the diagnosis with the given diagnosis id
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Diagnosis getDiagnosisById(Integer diagnosisId) {
+		return (Diagnosis) sessionFactory.getCurrentSession().get(Diagnosis.class, diagnosisId);
+	}
+
+
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Can you remove these empty lines?

**Comment by dkayiwa** [comment]
Can you also add some tests along these lines? https://github.com/openmrs/openmrs-module-emrapi/blob/9257dcd31d4d7af72742e78337662988b1edb463/condition-list/src/test/java/org/openmrs/module/emrapi/conditionslist/db/ConditionDAOTest.java

**Comment by paradisekelechi** [comment]
@CeciliaCaroline 
Kindly modify the PR title to match the convention with the ticket name starting it.

File: webapp/src/test/resources/override-web.xml [hunk 4]
```diff
@@ -10,10 +10,10 @@
     graphic logo is a trademark of OpenMRS Inc.
 
 -->
-<web-app xmlns="http://java.sun.com/xml/ns/j2ee"
-	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-	xsi:schemaLocation="http://java.sun.com/xml/ns/j2ee http://java.sun.com/xml/ns/j2ee/web-app_2_4.xsd"
-	version="2.4">
+<web-app xmlns="http://java.sun.com/xml/ns/javaee"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-Instance"
+		 xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
```
Comment (dkayiwa) [inline webapp/src/test/resources/override-web.xml]: Did you make any changes to this file?

File: webapp/src/main/webapp/WEB-INF/web.xml [hunk 5]
```diff
@@ -10,11 +10,11 @@
     graphic logo is a trademark of OpenMRS Inc.
 
 -->
-<web-app xmlns="http://java.sun.com/xml/ns/j2ee"
-	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-	xsi:schemaLocation="http://java.sun.com/xml/ns/j2ee http://java.sun.com/xml/ns/j2ee/web-app_3_0.xsd"
-	version="3.0">
-
+<web-app xmlns="http://java.sun.com/xml/ns/javaee"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-Instance"
```
Comment (dkayiwa) [inline webapp/src/main/webapp/WEB-INF/web.xml]: Did you make any changes to this file?

File: webapp/src/test/resources/override-web.xml (hunk 4 above)
Comment (CeciliaCaroline) [reply webapp/src/test/resources/override-web.xml]: Nope I just rebased

File: webapp/src/main/webapp/WEB-INF/web.xml (hunk 5 above)
Comment (CeciliaCaroline) [reply webapp/src/main/webapp/WEB-INF/web.xml]: No, I just rebased

Comment (dkayiwa) [reply webapp/src/main/webapp/WEB-INF/web.xml]: So can you remove the changes that i see in this pr?

Comment (CeciliaCaroline) [reply webapp/src/main/webapp/WEB-INF/web.xml]: OKay

Comment (CeciliaCaroline) [reply webapp/src/main/webapp/WEB-INF/web.xml]: When I remove the changes, it causes merge conflicts. Even when I remove them locally, they still persist.

Comment (CeciliaCaroline) [reply webapp/src/main/webapp/WEB-INF/web.xml]: Merge conflicts resolved

File: webapp/src/test/resources/override-web.xml [hunk 6]
```diff
@@ -10,10 +10,12 @@
     graphic logo is a trademark of OpenMRS Inc.
 
 -->
-<web-app xmlns="http://java.sun.com/xml/ns/javaee"
-		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-Instance"
-		 xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
-		 version="3.0">
+
+<web-app xmlns="http://java.sun.com/xml/ns/j2ee"
+		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (dkayiwa) [inline webapp/src/test/resources/override-web.xml]: Can you revert the changes to this file?

File: webapp/src/main/webapp/WEB-INF/web.xml [hunk 7]
```diff
@@ -10,9 +10,10 @@
     graphic logo is a trademark of OpenMRS Inc.
 
 -->
-<web-app xmlns="http://java.sun.com/xml/ns/javaee"
-		 xmlns:xsi="http://www.w3.org/2001/XMLSchema-Instance"
-		 xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
+
```
Comment (dkayiwa) [inline webapp/src/main/webapp/WEB-INF/web.xml]: Can you revert the changes to this file?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
+ * @since 2.2
+ */
+public class HibernateDiagnosisDAO implements DiagnosisDAO {
+	
+	private static final Log log = LogFactory.getLog(DiagnosisDAO.class);
```
[the remark this answers is not in the document]
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: The class name in the log factory is not correct

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+
+/**
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
+ * @since 2.2
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Since this class is not exposed to the API clients, you can remove the above @since annotation

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Can you change this to the latest license header file?

File: webapp/src/main/webapp/WEB-INF/web.xml (hunk 7 above)
Comment (dkayiwa) [reply webapp/src/main/webapp/WEB-INF/web.xml]: Did you see the above comment?

File: webapp/src/test/resources/override-web.xml (hunk 6 above)
Comment (CeciliaCaroline) [reply webapp/src/test/resources/override-web.xml]: I did revert but each time I rebase, the file is changed. It seems these changes have been made to master

File: webapp/src/main/webapp/WEB-INF/web.xml (hunk 7 above)
Comment (CeciliaCaroline) [reply webapp/src/main/webapp/WEB-INF/web.xml]: I did revert but each time I rebase, the file is changed. It seems these changes have been made to master

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 10 above)
Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: On it

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 9 above)
Comment (CeciliaCaroline) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Done

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 8 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: I think nowadays we prefer to have loggers as **protected final**

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	/**
+	 * Saves the diagnosis.
+	 *
+	 * @param diagnosis the diagnosis to save.
+	 * @return the saved diagnosis.
+	 */
+	@Override
+	@Transactional
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: @Transactional annotations need to be on the Service Implementation

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+	 *
+	 * @param patient the patient whose active diagnoses are being queried.
+	 * @return all active diagnoses associated with the specified patient.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	@Transactional(readOnly = true)
+	public List<Diagnosis> getActiveDiagnoses(Patient patient) {
+		Query query = sessionFactory.getCurrentSession().createQuery(
+			"from Diagnosis d where d.patient.patientId = :patientId and d.voided = false "
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: I'd imagine HQL should figure out that it needs to compare the database ids so d.patient = :patient would work

Comment (wluyima) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: What are active Diagnoses? I'd think getDiagnoses(Patient) by default returns non voided one

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+				+ "by d.dateCreated desc");
+		query.setInteger("patientId", patient.getId());
+		return query.list();
+	}
+
+	public List<Diagnosis> getDiagnoses(Encounter encounter){
+		return null;
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: What is this for?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+		updateSearchIndex();
+		
+	}
+
+	@Test
+	public void shouldDiagnosisByUuid() {
+		String uuid = "4e663d66-6b78-11e0-93c3-18a905e044dc";
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisByUuid(uuid);
+		assertEquals(diagnosis.getUuid(), uuid);
+	}
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: You should be testing the service layer methods that delegate to these

File: api/src/test/resources/org/openmrs/api/include/HibernateDiagnosisDAOTest-Dataset.xml [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    This Source Code Form is subject to the terms of the Mozilla Public License,
+    v. 2.0. If a copy of the MPL was not distributed with this file, You can
+    obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+    the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+    Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+    graphic logo is a trademark of OpenMRS Inc.
+-->
+
+<dataset>
+	<diagnosis diagnosis_id="1"
```
Comment (wluyima) [inline api/src/test/resources/org/openmrs/api/include/HibernateDiagnosisDAOTest-Dataset.xml]: Conventionally each row is on one line in the test data file

File: api/src/test/resources/org/openmrs/api/include/HibernateDiagnosisDAOTest-Dataset.xml [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+			   creator="1" 
+			   date_created="2015-01-01 00:00:00.0" 
+			   voided="false" 
+			   uuid="4e663d66-6b78-11e0-93c3-18a905e044dc"
+			   rank="1"
+			   condition_verification_status="PROVISIONAL"
+			   coded_or_free_text=""
+			  />
+	
+	<!--<diagnosis diagnosis_id="2" patient_id="2" creator="2" date_created="2015-01-02 00:00:00.0" voided="false" uuid="6f85b2a6-6b78-11e0-93c3-18a905e044dc" />-->
```
Comment (wluyima) [inline api/src/test/resources/org/openmrs/api/include/HibernateDiagnosisDAOTest-Dataset.xml]: If you are not using these, remove them rather than commenting them out

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
+ * @since 2.2
+ */
+public class HibernateDiagnosisDAO implements DiagnosisDAO {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Does the spring bean for this class already exist?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 8 above)
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: The Java best practice seems to advocate for private loggers. 😊

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 14 above)
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: I actually feel that testing should be closer to the functionality we are testing. For instance, since this is testing data access methods, then the DAO test is more appropriate than the service layer test. If the service layer adds more business logic, then we can also have the service layer test.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 11 above)
Comment (CeciliaCaroline) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: @transactional annotations removed

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 12 above)
Comment (CeciliaCaroline) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: My current implementation of getDiagnoses gets all diagnoses for a given encounter without consideration of whether they are voided or not and that is why there is the method to getActiveDiagnoses which returns non voided ones only

File: webapp/src/test/resources/override-web.xml (hunk 6 above)
Comment (dkayiwa) [reply webapp/src/test/resources/override-web.xml]: No merge has been done on master. If i file changes accidentally, you do not have to commit it. Can you ask your team mates for help? We just need a pull request which does not have changes to this file.

File: webapp/src/main/webapp/WEB-INF/web.xml (hunk 7 above)
Comment (dkayiwa) [reply webapp/src/main/webapp/WEB-INF/web.xml]: No merge has been done on master. If i file changes accidentally, you do not have to commit it. Can you ask your team mates for help? We just need a pull request which does not have changes to this file.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.openmrs.Diagnosis;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.test.BaseModuleContextSensitiveTest;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class HibernateDiagnosisDAOTest extends BaseModuleContextSensitiveTest {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: This is the cause of your error. The base class should be BaseContextSensitiveTest
The one that you have is only used in modules.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	
+	private static final String DIAGNOSIS_XML = "HibernateDiagnosisDAOTest-Dataset.xml";
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: This file will not be found. It needs the full path as here https://github.com/openmrs/openmrs-core/pull/2590/commits/d4981cd723b8dd55064c1b58ae6b0bcc4a19665f#diff-0d16b7abc7a5851414804ac8d1b34bb9R31

Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Done

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 18 above)
Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Noted

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.test.BaseContextSensitiveTest;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class HibernateDiagnosisDAOTest extends BaseContextSensitiveTest {
+	
+	private static final String DIAGNOSIS_XML = "org/openmrs/api/include/HibernateDiagnosisDAOTestDataset.xml";
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: What do you use this one for?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.Before;
+import org.junit.rules.ExpectedException;
+import org.openmrs.Diagnosis;
+import org.openmrs.api.db.DiagnosisDAO;
+import org.openmrs.test.BaseContextSensitiveTest;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class HibernateDiagnosisDAOTest extends BaseContextSensitiveTest {
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Can we add more tests to this class for the other methods?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+import org.openmrs.api.db.DiagnosisDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+
+/**
+ * Hibernate implementation of the DiagnosisDAO
+ *
+ * @see DiagnosisDAO
+ * @see org.openmrs.api.DiagnosisService
+ *
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Can you remove the above empty line?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 20 above)
Comment (dkayiwa) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Are you using the above for anything?

Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: I do not think so. Let me remove it.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 22 above)
Comment (CeciliaCaroline) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Done

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+	 * Gets all active diagnoses related to the specified patient.
+	 *
+	 * @param patient the patient whose active diagnoses are being queried.
+	 * @return all active diagnoses associated with the specified patient.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<Diagnosis> getActiveDiagnoses(Patient patient) {
+		Query query = sessionFactory.getCurrentSession().createQuery(
+//			"from Diagnosis d where d.patient.patientId = :patientId and d.voided = false "
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: You need to remove code rather than commenting it out unless you plan on adding it back

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+	 * @return all active diagnoses associated with the specified patient.
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<Diagnosis> getActiveDiagnoses(Patient patient) {
+		Query query = sessionFactory.getCurrentSession().createQuery(
+//			"from Diagnosis d where d.patient.patientId = :patientId and d.voided = false " 
+//				+ "order by d.dateCreated desc");
+			"from Diagnosis c where c.patient.patientId = :patientId and c.voided = false order "
+				+ "by c.onsetDate desc");
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Is it a requirement to have them ordered by onsetDate?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+	/**
+	 * Gets all diagnoses for a given encounter
+	 *
+	 * @param encounter the specific encounter to get the diagnoses for.
+	 * @return list of diagnoses for an encounter
+	 */
+	@Override
+	public List<Diagnosis> getDiagnoses(Encounter encounter){
+		Query query = sessionFactory.getCurrentSession().createQuery(
+			"from Diagnosis d where d.encounter = :encounterId order by d.dateCreated desc");
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: Is it a requirement to have them sorted by dateCreated? What's the purpose of onsetDate if you are using dateCreated here?

File: api/src/main/resources/org/openmrs/api/db/hibernate/Diagnosis.hbm.xml [hunk 26]
```diff
@@ -54,10 +57,10 @@
 		
 		<many-to-one name="changedBy" class="org.openmrs.User" column="changed_by" />
 
-		<many-to-one name="encounter" class="org.openmrs.Encounter"
+		<many-to-one name="encounter" class="org.openmrs.Encounter" column="encounter_id"
 					 not-null="true" />
 
-		<many-to-one name="condition" class="org.openmrs.Condition"/>
+		<many-to-one name="condition" class="org.openmrs.Condition" column="condition_id"/>
```
Comment (wluyima) [inline api/src/main/resources/org/openmrs/api/db/hibernate/Diagnosis.hbm.xml]: I see you've update the hbm file but I don't see the liquibase changesets to update the table in the database, are they already added?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	@Test
+	public void shouldDiagnosisById() {
+		int diagnosisId = 1;
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisById(diagnosisId);
+		assertEquals((int) diagnosis.getId(), diagnosisId);
+	}
+
+	//	@Test
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Why is this commented out?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+		String uuid = "4e663d66-6b78-11e0-93c3-18a905e044dc";
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisByUuid(uuid);
+		assertEquals(diagnosis.getUuid(), uuid);
+	}
+
+	@Test
+	public void shouldDiagnosisById() {
+		int diagnosisId = 1;
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisById(diagnosisId);
+		assertEquals((int) diagnosis.getId(), diagnosisId);
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: You should check the uuid instead

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+		String uuid = "4e663d66-6b78-11e0-93c3-18a905e044dc";
+		Diagnosis diagnosis = diagnosisDAO.getDiagnosisByUuid(uuid);
+
+		Assert.assertNotNull(diagnosis);
+
+		diagnosisDAO.deleteDiagnosis(diagnosis);
+
+		Assert.assertNull(diagnosisDAO.getDiagnosisByUuid(uuid));
+		}
+
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: I think you are missing several other tests for the other DAO methods?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java (hunk 23 above)
Comment (CeciliaCaroline) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAO.java]: noted

File: api/src/main/resources/org/openmrs/api/db/hibernate/Diagnosis.hbm.xml (hunk 26 above)
Comment (CeciliaCaroline) [reply api/src/main/resources/org/openmrs/api/db/hibernate/Diagnosis.hbm.xml]: Yes, they are already added

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 29 above)
Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: More tests are being added

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 27 above)
Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: This test is currently work in progress which is why it is commented out.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java (hunk 28 above)
Comment (CeciliaCaroline) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernateDiagnosisDAOTest.java]: Working on it

**Comment by dkayiwa** [comment]
Closing this because of https://github.com/openmrs/openmrs-core/pull/2612
